### ISSUE TITLE
Google Trace degrades performance of our application

### ISSUE THREAD

**Issue by nover** [googleapis/google-cloud-dotnet#5966]
This is a cry for help / help us understand what we're doing wrong kind of issue :)

We have added Google Trace to a series of APIs for a client to help investigate some performance problems, however, on one of the projects, adding Google Trace just made the performance problems worse. 

I have managed to narrow it down to a very simple repro (I can share the repository if you need it) which shows that when we add the following `TraceFallbackPredicate` things go south. It causes about a 10x slowdown in request handling capabilities on my local test-bed hardware, and we're seeing similar degradation in our production deploy:
```
options.TraceFallbackPredicate = TraceDecisionPredicate.Create(request =>
{
    // Do not trace OPTIONS 
    var isOptionsCall = request.Method.ToLowerInvariant().Equals("options");
                
    // Do not trace our monitoring routes
    var isMonitoringRoute =
        request.Path.Equals(PathString.FromUriComponent("/"));
                
    return !(isOptionsCall || isMonitoringRoute);
});
```

Basically filter out `OPTIONS` calls and health checks on `/`.

I have performed some profiling with Rider, the first screenshot shows the flamegraph WITH our custom predicate - a lot of time is spent in `EndSpan`:
![Markup on 2021-02-23 at 13:08:21](https://user-images.githubusercontent.com/866244/108844832-e7f85a80-75dc-11eb-938c-d7b2251f55c5.png)

The following screenshot is with the predicate removed, which looks more sane:
![Markup on 2021-02-23 at 13:11:07](https://user-images.githubusercontent.com/866244/108844925-08281980-75dd-11eb-9d22-4bb3b5d520ad.png)

The code profiled does some network IO and returns a response.

So what are we doing wrong? Can that trace predicate really cause this, or is something else at play?

**Comment by nover** [comment]
reproduction repository is available here: https://github.com/nover/google-trace-performance-degradation-repro

**Comment by amanda-tarafa** [comment]
I'll take a look at this in the afternoon and come back as soon as I know more.

**Comment by amanda-tarafa** [comment]
By adding the trace predicate you are effectively tracing almost every request (except for the ones you are filtering out) as oposed to tracing 1 request per second. That's not a problem in itself, but when you combine that with the fact that you are using no buffer, then you get the latency increase you are seing (see #5632 for instance and also #5275 which is tangentially related), because now you have one "extra" service call per request. My suggestion is simply to use buffers. If you are using auto-scalable services, like Google's Cloud Run, you cannot use the timed buffer because these environments usually don't guarantee CPU cycles during downtime, in those cases you should use the sized buffer. Else you can use either buffering strategy.

I've looked at making `ISpan` to be `IAsyncDisposable` before, which would help mitigate some of this but it would involve some major and even breaking changes, and we've felt that the amount of work involved does not makes up for just a mitigation, you can still have enough requests so that the change doesn't help much. And, using buffers is mitigation enough, and a natural strategy to use for monitoring, and that you can even tweak over time depending on your load.

Let me know if this works for you.

**Comment by nover** [comment]
Thanks for you always detailed answers @amanda-tarafa - I'll try to look into a combination of buffers and QPS limiting.

Side note : the thing about not using buffers is basically because we just went with what it's the getting started docs. Might it be worth it to note someting in the docs about implications of not using buffers? (Step 4)
https://cloud.google.com/trace/docs/setup/aspnetcore#configuring_the_client_library

**Comment by amanda-tarafa** [comment]
I'll take a look at the docs tomorrow to see how best we can clarify.

**Comment by nover** [comment]
@amanda-tarafa I can confirm that enabling the `BufferOptions.TimedBuffer` and setting a `qpsSampleRate` of `1` in the repro brings the speed back to how it is without Google Trace added.

Feel free to close the issue.

**Comment by amanda-tarafa** [comment]
Thanks for confirming! You can probably set QPS higher than 1 now if you need to see more traces.

I'll close this one, but I've created GoogleCloudPlatform/dotnet-docs-samples#1266 to look into clarifying docs.

Thanks!

**Comment by nover** [comment]
> Thanks for confirming! You can probably set QPS higher than 1 now if you need to see more traces.

Yes I can indeed, reran my test with `QPS:10` and `QPS:100` without any noticeable performance degradation (test suite is generating ~800 req/s)
